internal/cmd/local/docker/docker.go

```diff
@@ -153,3 +175,205 @@ func (d *Docker) Port(ctx context.Context, container string) (int, error) {
+}
+
+// volumeExists returns true if the volumeID exists on the host machine, false otherwise.
+func (d *Docker) volumeExists(ctx context.Context, volumeID string) string {
+	if v, err := d.Client.VolumeInspect(ctx, volumeID); err != nil {
+		pterm.Debug.Println(fmt.Sprintf("Volume %s cannot be accessed: %s", volumeID, err))
+		return ""
+	} else {
+		return v.Mountpoint
+	}
+}
+
+// Exec executes an exec cmd against the container.
+// Largely inspired by the official docker client - https://github.com/docker/cli/blob/d69d501f699efb0cc1f16274e368e09ef8927840/cli/command/container/exec.go#L93
+func (d *Docker) exec(ctx context.Context, container string, cmd []string) error {
+	if _, err := d.Client.ContainerInspect(ctx, container); err != nil {
+		return fmt.Errorf("could not inspect container '%s': %w", container, err)
+	}
+
+	resCreate, err := d.Client.ContainerExecCreate(ctx, container, types.ExecConfig{Cmd: cmd})
+	if err != nil {
+		return fmt.Errorf("could not create exec for container '%s': %w", container, err)
+	}
+
+	if err := d.Client.ContainerExecStart(ctx, resCreate.ID, types.ExecStartCheck{}); err != nil {
+		return fmt.Errorf("could not start exec for container '%s': %w", container, err)
+	}
+
+	ticker := time.NewTicker(500 * time.Millisecond) // how often to check
+	timer := time.After(5 * time.Minute)             // how long to wait
+	running := true
+
+	// loop until the exec command returns a "Running == false" status, or until we've hit our timer
+	for running {
+		select {
+		case <-ticker.C:
+			res, err := d.Client.ContainerExecInspect(ctx, resCreate.ID)
+			if err != nil {
+				return fmt.Errorf("could not inspect container '%s': %w", container, err)
+			}
+			running = res.Running
+			if res.ExitCode != 0 {
+				return fmt.Errorf("container '%s' exec exited with non-zero exit code: %d", container, res.ExitCode)
+			}
+		case <-timer:
+			return errors.New("timed out waiting for docker exec to complete")
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+
+	return nil
+}
+
+// copyFromContainer emulates the `docker cp` command.
+// The dst will be treated as a directory.
+func (d *Docker) copyFromContainer(ctx context.Context, container, src, dst string) error {
+	// ensure dst directory exists
+	if err := os.MkdirAll(path.Dir(dst), 0755); err != nil {
+		return fmt.Errorf("could not create directory '%s': %w", dst, err)
+	}
+
+	reader, stat, err := d.Client.CopyFromContainer(ctx, container, src)
+	if err != nil {
+		return fmt.Errorf("could not copy from container '%s': %w", container, err)
+	}
+	defer reader.Close()
+
+	copyInfo := archive.CopyInfo{
+		Path:   src,
+		Exists: true,
+		IsDir:  stat.Mode.IsDir(),
+	}
+
+	if err := archive.CopyTo(reader, copyInfo, dst); err != nil {
+		return fmt.Errorf("could not copy from container '%s': %w", container, err)
+	}
+
+	return nil
+}
+
+// stopContainer will stop and ultimately remove the containerID
+func (d *Docker) stopContainer(ctx context.Context, containerID string) {
```

Comment:
suggestion: Let's add a flag here to control the removal behavior, right now it is not obvious from just the function signature/name that its going to do that. And I would not expect it, by default

Reply:
I'll rename the method instead to `stopAndRemove...` as there is currently not need to only stop a container without removing it.